# Linkage attack

In [25]:
from sqlalchemy import create_engine, MetaData, Table, Column, Numeric, Integer, VARCHAR, text
from sqlalchemy.orm import scoped_session, sessionmaker
import pandas as pd


engine = create_engine("postgresql://postgres:password@localhost:5432/pds_proj_1")
CONN = engine.connect()
META_DATA = MetaData()
META_DATA.reflect(bind=engine)

MED_TABLE = META_DATA.tables["med_data"]
WORK_TABLE = META_DATA.tables["work_data"]

with engine.connect() as conn:
    result = conn.execute(MED_TABLE.select())
    MED_TABLE = result.fetchall()
    
    result = conn.execute(WORK_TABLE.select())
    WORK_TABLE = result.fetchall()


med_df = pd.DataFrame(MED_TABLE)
work_df = pd.DataFrame(WORK_TABLE)

In [26]:
med_df.head()

id      age  gender postal_code diagnosis
0  b1f8077f-69bc-4ac0-bcaf-ca6e617901c9  20 - 24    Male   4967700**   V9107XA
1  3da6e880-782c-4692-99f9-50186a76ac96  20 - 24  Female   2072119**   T82328A
2  50388100-f445-4dcb-9f8d-0ea8f2114f10  25 - 29    Male   9393034**   S92403K
3  0fd1e26f-9b6c-4ec8-b32b-ec7fc45c4168  15 - 19  Female   2952691**     C8197
4  778b3ffb-c09c-47f1-a298-0a04ac0844e7  20 - 24    Male   4206415**   V9206XA

In [27]:
work_df.head()

id    f_name      l_name postal_code  gender  \
0   1   Tamqrah  Shawdforth   133261301  Female   
1   2     Ardys    Ferenczi       64744  Female   
2   3  Margette      Dibner       62606  Female   
3   4     Libby    Gandrich   593301943  Female   
4   5   Lennard    Castelin   006311003    Male   

                           education          workplace  \
0     Fachhochschule Bonn-Rhein-Sieg  Schamberger Group   
1  Chaudhary Charan Singh University            Gutmann   
2              University of Sudbury              Doyle   
3            University College Cork        Jenkins Inc   
4         University of West Hungary   VonRueden-Dooley   

                department  
0       Product Management  
1     Cummings and Cormier  
2   Powlowski and Leuschke  
3                 Training  
4                Marketing

In [ ]:
# useless for now
def degeneralize_age(age_range):
    age_lower, age_higher = age_range.split(" - ")
    return int(age_lower), int(age_higher)

In [19]:
def suppress_post_code(code, supp_range = 2):
    post_code_supped = list(str(code))
    post_code_supped[-supp_range:] = ["*"]*supp_range
    post_code_supped = ''.join(post_code_supped)
    return post_code_supped

In [ ]:
work_df['postal_code'] = work_df['postal_code'].apply(suppress_post_code)

In [29]:
work_df.head()

id    f_name      l_name postal_code  gender  \
0   1   Tamqrah  Shawdforth   1332613**  Female   
1   2     Ardys    Ferenczi       647**  Female   
2   3  Margette      Dibner       626**  Female   
3   4     Libby    Gandrich   5933019**  Female   
4   5   Lennard    Castelin   0063110**    Male   

                           education          workplace  \
0     Fachhochschule Bonn-Rhein-Sieg  Schamberger Group   
1  Chaudhary Charan Singh University            Gutmann   
2              University of Sudbury              Doyle   
3            University College Cork        Jenkins Inc   
4         University of West Hungary   VonRueden-Dooley   

                department  
0       Product Management  
1     Cummings and Cormier  
2   Powlowski and Leuschke  
3                 Training  
4                Marketing

In [30]:
matches = pd.merge(
    work_df, 
    med_df, 
    on=['postal_code', 'gender'],
    how='inner',
    suffixes=('_work', '_med')
)

In [ ]:
print(f"Found {len(matches)} exact matches on postal_code + gender")

print("\nRe-identified individuals and their diagnoses:")
for _, match in matches.head(10).iterrows():
    print(f"{match['f_name']} {match['l_name']} -> Diagnosis: {match['diagnosis']}")

Found 830 exact matches on postal_code + gender

Re-identified individuals and their diagnoses:
Tamqrah Shawdforth -> Diagnosis: S52372Q
Ardys Ferenczi -> Diagnosis: T343XXS
Libby Gandrich -> Diagnosis: I63212
Lennard Castelin -> Diagnosis: S60476
Phil Abelov -> Diagnosis: T3370XA
Susy Reuther -> Diagnosis: M14849
Vania Klousner -> Diagnosis: L940
Tedda Whotton -> Diagnosis: T22292S
Francisco Ettels -> Diagnosis: V9388XD
Marlie Patillo -> Diagnosis: T508X3


For postal code suppresion =2 the linkage attack managed to re-identify 830 individuals. Which is very successful